inference/enterprise/workflows/execution_engine/compiler/graph_constructor.py

```diff
@@ -0,0 +1,602 @@
+        if node_data.get("kind") != STEP_NODE_KIND:
+            reversed_steps_graph.remove_node(node)
+    return reversed_steps_graph
+
+
+def verify_multi_parent_step_execution_paths(
+    reversed_steps_graph: nx.DiGraph,
+    reversed_topological_order: List[str],
+    step: str,
+) -> None:
+    control_flow_steps_successors = defaultdict(set)
+    max_conditions_steps_in_execution_branch = 0
+    for parent_of_investigated_step in reversed_steps_graph.successors(step):  # O(V)
+        reversed_flow_path = (
+            construct_path_to_step_through_selected_parent(  # O(E) -> O(V^2)
+                graph=reversed_steps_graph,
+                topological_order=reversed_topological_order,
+                parent_step=parent_of_investigated_step,
+                step=step,
+            )
+        )
+        (
+            control_flow_steps_successors,
+            condition_steps,
+        ) = denote_flow_control_steps_successors_in_normal_flow(  # O(V)
+            reversed_steps_graph=reversed_steps_graph,
+            reversed_flow_path=reversed_flow_path,
+            control_flow_steps_successors=control_flow_steps_successors,
+        )
+        max_conditions_steps_in_execution_branch = max(
+            condition_steps, max_conditions_steps_in_execution_branch
+        )
+    if len(control_flow_steps_successors) > max_conditions_steps_in_execution_branch:
+        raise ConditionalBranchesCollapseError(
+            public_message=f"In execution graph, detected collision of branches that originate "
+            f"in different flow-control steps. When using flow control step you create "
+            f"a sub-workflow, which cannot reach any step from the outside.",
+            context="workflow_compilation | execution_graph_construction",
+        )
+    for condition_step, potential_next_steps in control_flow_steps_successors.items():
+        if len(potential_next_steps) > 1:
+            raise ConditionalBranchesCollapseError(
+                public_message=f"In execution graph, in flow-control step: {condition_step} there are originated "
+                f"workflow branches that clashes together.",
+                context="workflow_compilation | execution_graph_construction",
+            )
+
+
+def construct_path_to_step_through_selected_parent(
+    graph: nx.DiGraph,
+    topological_order: List[str],
+    parent_step: str,
+    step: str,
+) -> List[str]:
+    normal_flow_path_nodes = nx.descendants(graph, parent_step)
+    normal_flow_path_nodes.add(parent_step)
+    normal_flow_path_nodes.add(step)
+    return [n for n in topological_order if n in normal_flow_path_nodes]
+
+
+def denote_flow_control_steps_successors_in_normal_flow(
+    reversed_steps_graph: nx.DiGraph,
+    reversed_flow_path: List[str],
+    control_flow_steps_successors: Dict[str, Set[str]],
+) -> Tuple[Dict[str, Set[str]], int]:
+    conditions_steps = 0
+    if len(reversed_flow_path) == 0:
```

Comment:
`if len(reversed_flow_path) == 0:` -> `if not reversed_flow_path:`

Reply:
👍